File: js/test/msea/unionApi.test.ts
import process from 'process';

import { MapleStoryApiError, MapleStoryApiErrorCode } from '../../src';
import { MapleStoryApi } from '../../src/maplestory/api/msea';
import { toString } from '../utils';

const apiKey = process.env.API_KEY_MSEA!; // Your API Key
const api = new MapleStoryApi(apiKey);

const ocid = 'caee7cf6ec6cb067c33b899f2779f3d1';

describe('Union Information Retrieval', () => {
  describe('getUnion', () => {
    test('success: getUnion', async () => {
      const response = await api.getUnion(ocid);
      // nothing to assert because some characters may not have union
      console.log(toString(response));
    });

    test('fail: getUnion with invalid date', async () => {
      try {
        await api.getUnion(ocid, {
          year: 2025,
          month: 4,
          day: 19,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2025-04-20.');
        console.log(error.message);
      }
    });

    test('fail: getUnion with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getUnion(invalidOcid);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getUnionRaider', () => {
    test('success: getUnionRaider', async () => {
      const response = await api.getUnionRaider(ocid);
      // nothing to assert because some characters may not have union raider
      console.log(toString(response));
    });

    test('fail: getUnionRaider with invalid date', async () => {
      try {
        await api.getUnionRaider(ocid, {
          year: 2025,
          month: 4,
          day: 19,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2025-04-20.');
        console.log(error.message);
      }
    });

    test('fail: getUnionRaider with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getUnionRaider(invalidOcid);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getUnionArtifact', () => {
    test('success: getUnionArtifact', async () => {
      const response = await api.getUnionArtifact(ocid);
      // nothing to assert because some characters may not have union artifact
      console.log(toString(response));
    });

    test('fail: getUnionArtifact with invalid date', async () => {
      try {
        await api.getUnionArtifact(ocid, {
          year: 2025,
          month: 4,
          day: 19,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2025-04-20.');
        console.log(error.message);
      }
    });

    test('fail: getUnionArtifact with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getUnionArtifact(invalidOcid);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });

  describe('getUnionChampion', () => {
    test('success: getUnionChampion', async () => {
      const response = await api.getUnionChampion(ocid);
      // nothing to assert because some characters may not have union champion
      console.log(toString(response));
    });

    test('success: getUnionChampion with date', async () => {
      const response = await api.getUnionChampion(ocid, {
        year: 2025,
        month: 12,
        day: 19,
      })
      expect(response).toBeDefined()
      console.log(toString(response))
    })

    test('fail: getUnionChampion with invalid date', async () => {
      try {
        await api.getUnionChampion(ocid, {
          year: 2025,
          month: 12,
          day: 17,
        });
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as Error;
        expect(error).toBeInstanceOf(Error);
        expect(error.message).toContain('You can only retrieve data after 2025-12-18.');
        console.log(error.message);
      }
    });

    test('fail: getUnionChampion with invalid ocid throw OPENAPI00003', async () => {
      const invalidOcid = 'invalid_ocid_123';
      try {
        await api.getUnionChampion(invalidOcid);
        fail('An error should have been thrown.');
      } catch (e) {
        const error = e as MapleStoryApiError;
        expect(error).toBeInstanceOf(MapleStoryApiError);
        expect(error.errorCode).toBe(MapleStoryApiErrorCode.OPENAPI00003);
        console.log(error.errorCode, error.message);
      }
    });
  });
});

File: python/test/msea/union_api_test.py
import os
import sys
import unittest
import pytest

from datetime import datetime
from dotenv import load_dotenv

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../')))

from maplestory_openapi.api.common import MapleStoryApiException
from maplestory_openapi.api.msea import MapleStoryApi

load_dotenv()

api_key = os.environ['API_KEY_MSEA'] # Your API Key
api = MapleStoryApi(api_key)

ocid = 'caee7cf6ec6cb067c33b899f2779f3d1'

# Union Information Retrieval

class TestGetUnion(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_union(self):
        response = await api.get_union(ocid)
        # nothing to assert because some characters may not have union
        print(response)

    async def test_success_get_union_with_date(self):
        response = await api.get_union(ocid, date=datetime(2025, 4, 20))
        assert response is not None
        print(response)

    async def test_fail_get_union_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_union(ocid, date=datetime(2023, 4, 19))
        assert 'You can only retrieve data after 2025-04-20' in str(e.value)
        print(e.value)

    async def test_fail_get_union_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_union(invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


class TestGetUnionRaider(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_union_raider(self):
        response = await api.get_union_raider(ocid)
        # nothing to assert because some characters may not have union raider
        print(response)

    async def test_success_get_union_raider_with_date(self):
        response = await api.get_union_raider(ocid, date=datetime(2025, 4, 20))
        assert response is not None
        print(response)

    async def test_fail_get_union_raider_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_union_raider(ocid, date=datetime(2023, 4, 19))
        assert 'You can only retrieve data after 2025-04-20' in str(e.value)
        print(e.value)

    async def test_fail_get_union_raider_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_union_raider(invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


class TestGetUnionArtifact(unittest.IsolatedAsyncioTestCase):
    async def test_success_get_union_artifact(self):
        response = await api.get_union_artifact(ocid)
        # nothing to assert because some characters may not have union artifact
        print(response)

    async def test_success_get_union_artifact_with_date(self):
        response = await api.get_union_artifact(ocid, date=datetime(2025, 4, 20))
        assert response is not None
        print(response)

    async def test_fail_get_union_artifact_with_invalid_date(self):
        with pytest.raises(Exception) as e:
            await api.get_union_artifact(ocid, date=datetime(2023, 4, 19))
        assert 'You can only retrieve data after 2025-04-20' in str(e.value)
        print(e.value)

    async def test_fail_get_union_artifact_with_invalid_ocid_throw_OPENAPI00003(self):
        invalid_ocid = 'invalid_ocid_123'
        with pytest.raises(MapleStoryApiException) as e:
            await api.get_union_artifact(invalid_ocid)
        assert e.value.error_code == 'OPENAPI00003'
        print(e.value.error_code, e.value.message)


if __name__ == '__main__':
    unittest.main()
